### ISSUE TITLE
How to avoid storing BLOB's in the offline reducer ?

### ISSUE THREAD

**Issue by clems36** [redux-offline/redux-offline#291]
### Are you using `offline()` or `createOffline()` API?
`offline()`

### Store config
```js
export const store = createStore(
  rootReducer,
  composeWithDevTools(
    applyMiddleware(reduxThunk),
    offline(offlineConfig),
  ),
);
```

### Offline config
```js
const offlineConfig = {
  ...defaultConfig,
  effect: async (effect, action) => {
    if (action.type === 'ADD_ATTACHMENT_REQUEST') {
      const newEffect = { ...effect };
      const { config: { data: { attachment } } } = newEffect;
      attachment.file = await fs.read(attachment.name, C.ENTITY_TYPE_ATTACHMENT);
      return new Promise(async (resolve: Function, reject: Function) => {
        try {
          let response = await axios(newEffect.config);
          response.config.data = null;
          console.log('response', response);
          resolve(response);
        } catch (error) {
          reject(error);
        }
      });
    } else {
      return axios(effect.config);
    }
  },
  queue: {
    ...defaultQueue,
    enqueue(outbox, query) {
      return [...outbox, query]
    },
    dequeue(outbox) {
      console.log('outbox', outbox);
      const [, ...rest] = outbox;
      return rest;
    },
  },
  persistCallback: () => {
    console.log('Rehydration complete');
    DeviceEventEmitter.emit('rehydrationEvent', { rehydrated: true });
  },
};
```
### The issue

Hi! I'm calling a webservice called `addAttachment` which receives a base64 string as param. 

If i were to use the normal flow of redux-offline, the base64 would get stored in the offline reducer for as long as the request hasn't been dequeued. But if my base64 is too long (a file of around 4 Mb on my device), i'm getting ```Error: database or disk is full (code 13)``` and the app crashes.

What I want to do is to avoid storing the base64 using redux. For that i'm using the `FileSystem` API from Expo to store the file there instead and just retrieve it in the `effect` before sending it to the webservice. Once the `axios` call is made, i'm setting the data to `null` before resolving the call's response to go on with the flow (i know, this is not a beautiful way to do it but i'm just experimenting right now).

Now when I log the `outbox` in the `dequeue`, the `effect`'s `config` still contains the base64. So my little trick inside the `effect` didn't change anything. I thought that the `dequeue`'s arguments were fed by the `effect`'s return object but I guess i'm wrong. 

Where does the `dequeue`'s `outbox` comes from ? How can I achieve what I'm trying to do ? 
Thanks!

**Comment by calumpeak** [comment]
Hi @clems36 
First off, I like the solution to the problem, really nice! I can't see any inherent problems with what is above.

When you say the "effects config still contains base64" do you mean that you're still seeing the base64 in the outbox queue? If that's the case have you definitely cleared all your indexdb caches as it might be that you're getting an old version rehydrating into your app when you reload.

Also double check that your offline action doesn't contain the base64 version anymore as it could also be that.

**Comment by clems36** [comment]
> When you say the "effects config still contains base64" do you mean that you're still seeing the base64 in the outbox queue?

Yes that is what I mean. Here's how the `config` of the `axios` response look like before resolving it:
``` js
"config": Object {
    "adapter": [Function xhrAdapter],
    "data": null,
    "headers": Object {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/json;charset=utf-8",
    },
    "maxContentLength": -1,
    "method": "post",
    "timeout": 0,
    "transformRequest": Object {
      "0": [Function transformRequest],
    },
    "transformResponse": Object {
      "0": [Function transformResponse],
    },
```
We can see that `data` is set to `null`. But then here's how the outbox looks like in the dequeue: 

``` js
"effect": Object {
          "config": Object {
            "data": Object {
              "attachment": Object {
                "entityLinks": Array [
                  Object {
                    "ids": Array [
                      9314290,
                    ],
                    "type": "JOB",
                  },
                ],
                "extId": "attached_from_mobile_1546425898382",
                "file": "/9j/4AAQSkZJRgABAQAAAQABAAD/2 etc...",
                "mimeType": "*/*",
                "name": "IMG-20190102-WA0001.jpg",
                "size": 99223,
                "url": "file:///data/user/0/host.exp.exponent/files/ExperienceData/%2540emixis%252Faidoo/attachment",
              },
              "sessionId": "4d2da3fd-b97a-46d5-90bb-0ec5973780ac",
            },
            "method": "POST",
            "url": "https://...",
          },
        },
```
Here I would expect `data` to be null as well but it has the initial value.

> have you definitely cleared all your indexdb caches

Can you please elaborate ? 

> check that your offline action doesn't contain the base64 version anymore

Do you mean in the effect (because in the dequeue that would be too late anyway) ? If so, would it really change anything to set the action's base64 to null there since we are not returning it ?

**Comment by calumpeak** [comment]
Are you seeing it in the outbox, or in an action? It's hard to tell.

> Can you please elaborate ?

Clear your application storage/cache to make sure you don't have any old entities in your outbox that gets rehydrated between refreshes. It could be that you have old data that you're testing, not new.

> expect data to be null

Try logging it before you resolve it to see what you're getting/checking the redux inspector.

**Comment by clems36** [comment]
> Are you seeing it in the outbox, or in an action? 

In the outbox.

> Clear your application storage/cache

I don't know how to do that but i'm pretty sure that the problem does not come from that. Before sending my action to redux, the outbox is empty, then when i send it, i clearly see that the request inside the `dequeue` that contains the base64 corresponds to the one where i removed it in the `effect`. Maybe i'm missing something here...

> what you're getting/checking the redux inspector

Unfortunately i can't get the remote dev tools to work on my project... which is why i'm referring to `console.log`. You can see in the original post where i'm placing them (`effect` and `dequeue`).

Basically what i'm asking is: shouldn't the outbox inside the `dequeue` reflect what is returned in the `effect` ?

Thanks for your help!

**Comment by calumpeak** [comment]
The outbox never contains what has been returned by effect. It only contains the original action containing the `meta.offline` payload. Dequeue will receive the `commit` or `rollback` action however to know to remove the first item in the queue.

**Comment by clems36** [comment]
Right, and dequeue will also receive the `effect` which in my case contains the base64. So I somehow have to set `meta.offline.effect.config.data.attachment.base64` to null before it gets stored through the offline reducer. But I can't do that right before dispatching my action because then the webservice call would not work... 

Here's how my action looks like:

```js
const addAttachmentRequest: Function = (config: IAddAttachmentConfig, attachment: IAttachment): Function => {
    console.log(ADD_ATTACHMENT_REQUEST);
    const requestId: number = Date.now();

    return (dispatch: Function): void => {
        dispatch({
            mandatory: true,
            meta: {
                offline: {
                    commit: { type: ADD_ATTACHMENT_SUCCESS, meta: { type: 'parent', localId: attachment.id, requestId } },
                    effect: { config },
                    rollback: { type: ADD_ATTACHMENT_ERROR, meta: { type: 'parent', localId: attachment.id, requestId } },
                },
            },
            payload: {
                attachment,
                isAddingAttachment: true,
                requestId: requestId,
            },
            type: ADD_ATTACHMENT_REQUEST,
        });
    };
};

export const addAttachment: Function = (sessionId: string, url: string, suffix: string, attachment: IAttachment): Function => {

    const config: IAddAttachmentConfig = {
        data: {
            attachment,
            sessionId,
        },
        method: 'POST',
        url: url + suffix,
    };

    return addAttachmentRequest(config, attachment);
};
```

**Comment by calumpeak** [comment]
So perhaps the `attachment` needs to be changed within the action. Only send:

```typescript
 const config: IAddAttachmentConfig = {
        data: {
            attachment: {
              name: attachment.name
            },
            sessionId,
        },
        method: 'POST',
        url: url + suffix,
    };
```
That will stop the entire blob from being passed in the action and stored should it not? (If I've understood the code correctly)

**Comment by clems36** [comment]
Of course but the webservice I'm using requires the attachment to have a base64 :)

**Comment by calumpeak** [comment]
Then I don't think there's a way to avoid it unless you want to store the base64 in another bit of local storage and reference it by ID in the effect. But it doesn't change the issue either way as you'd have to store/cleanup of it. That can be harder to reason about in the long run too. If your endpoint needs the base64 blob, it has to be stored and sent from somewhere.  Due to browser constraints you can't request the file data later as the file API's won't allow it, and the file could have been moved in the meantime on the users end even if you could do that which again, causes more issues.

Your server needs the file data one way or another, and if the application definitely needs this to be able to happen offline (for things like this I tend to restrict uploading to online only) then you're a bit stuck.

**Comment by clems36** [comment]
Thank you for clarifying this. Makes more sense now. I guess uploading big files on an offline smartphone is just something that I should reconsider. I might restrict this specific action to online only as you suggested, or just drastically restrict the size of the file...

Thanks for your time either way!

**Comment by calumpeak** [comment]
@clems36 Glad I could help :)

**Comment by Cattari** [comment]
Wanna share my experience here within a React Native app.

We stored the local URI in the action.
And during the sync, we used a specific matcher to determine what the request is.
In the matcher, we used the form data and changed the data sent to the server.

After the sync request is made, we clear the file using react-native-fs.

Works okay.
